// @undertheground/color/dist/color.js
export var black = "#000000";
export var white = "#ffffff";
export var pink99100 = "##FF0385";
export var pink9994 = "#F0027D";
export var pink9988 = "#E00275";
export var pink9982 = "#D1036F";
export var pink9976 = "#C20265";
export var pink9970 = "#B3025D";
export var pink9964 = "#A30255";
export var pink9958 = "#94014D";
export var pink9952 = "#94014D";
export var pink9946 = "#75013D";
export var pink9940 = "#660135";
export var pink9934 = "#57012D";
export var blue86100 = "#2478FF";
export var blue8694 = "#2271F0";
export var blue8688 = "#5584D2";
export var blue8682 = "#3970CA";
export var blue8676 = "#1C5BC2";
export var blue8670 = "#1954B3";
export var blue8664 = "#174DA3";
export var blue8658 = "#154594";
export var blue8652 = "#133E85";
export var blue8646 = "#103775";
export var blue8640 = "#0E3066";
export var blue8634 = "#0C2957";
export var grey91 = "#E8E8E8";
export var grey82 = "#D1D1D1";
export var grey73 = "#BABABA";
export var grey64 = "#A3A3A3";
export var grey55 = "#8C8C8C";
export var grey46 = "#757575";
export var grey37 = "#5E5E5E";
export var grey28 = "#5E5E5E";
export var grey19 = "#303030";
export var grey10 = "#1A1A1A";
var pink = {
    0: pink99100,
    1: pink9994,
    2: pink9988,
    3: pink9982,
    4: pink9976,
    5: pink9970,
    6: pink9964,
    7: pink9958,
    8: pink9952,
    9: pink9946,
    10: pink9940,
    11: pink9934
};
var blue = {
    0: blue86100,
    1: blue8694,
    2: blue8688,
    3: blue8682,
    4: blue8676,
    5: blue8670,
    6: blue8664,
    7: blue8658,
    8: blue8652,
    9: blue8646,
    10: blue8640,
    11: blue8634
};
var grey = {
    0: grey91,
    1: grey82,
    2: grey73,
    3: grey64,
    4: grey55,
    5: grey46,
    6: grey37,
    7: grey28,
    8: grey19,
    9: grey10
};
export var colors = {
    pink: pink,
    grey: grey,
    white: white,
    black: black,
    blue: blue
};
